Zend/zend_compile.c

```diff
@@ -1196,13 +1196,23 @@ zend_string *zend_type_to_string_resolved(zend_type type, zend_class_entry *scop
 			if (ZEND_TYPE_HAS_CE(*list_type)) {
 				str = add_type_string(str, ZEND_TYPE_CE(*list_type)->name);
 			} else {
-				zend_string *resolved = resolve_class_name(ZEND_TYPE_NAME(*list_type), scope);
-				str = add_type_string(str, resolved);
-				zend_string_release(resolved);
+				if (ZEND_TYPE_HAS_CE_CACHE(*list_type)
+				 && ZEND_TYPE_CE_CACHE(*list_type)) {
+					str = add_type_string(str, ZEND_TYPE_CE_CACHE(*list_type)->name);
+				} else {
+					zend_string *resolved = resolve_class_name(ZEND_TYPE_NAME(*list_type), scope);
+					str = add_type_string(str, resolved);
+					zend_string_release(resolved);
+				}
```

Comment:
I think with this change, it should be possible to remove the TYPE_CE kind completely. I don't think it was used for anything but property types.